step-api/src/main/java/org/jenkinsci/plugins/workflow/structs/DescribableHelper.java

```diff
@@ -323,14 +342,24 @@ static ParameterType of(Type type) {
                         for (Map.Entry<String,List<Class<?>>> entry : subtypesBySimpleName.entrySet()) {
                             if (entry.getValue().size() == 1) { // normal case: unambiguous via simple name
                                 try {
-                                    types.put(entry.getKey(), schemaFor(entry.getValue().get(0)));
+                                    String key = entry.getKey();
```

Comment:
I guess there is a bit of confusion here about what the values in the stack represent—in some cases `simpleName`s, in others FQNs. Probably it does not really matter: at worst you could recurse once and add both to the stack before stopping.